### ISSUE TITLE
"toggle last modified column" does not work for all files

### ISSUE THREAD

**Issue by MountComb** [jupyterlab/jupyterlab#8860]
## Description

Last modified column reappears when navigate to a directory containing more files.

If toggle last modified column in a folder containing 2 files and then create another 3 files I get:
![screenshot](https://user-images.githubusercontent.com/69704258/90712327-5af2e280-e2d5-11ea-9d35-0972e953282e.png)

May be related to #8258?

## Reproduce

1. Go to any directory A in the file browser.
2. Select "toggle last modified column" in the context menu of the file browser heading.
3. Go to a new directory B containing more files, or just create a new file in the same directory.

Now only the first X files hides the last modified column, where X is number of files in directory A.

## Expected behavior

The file browser should hide the last modified column for all files in directory B.

## Context
- Operating System and version: CentOS 6.8
- Browser and version: Firefox 79.0 & Chrome 84.0.4147.125
- JupyterLab version:
```console
$ jupyter --version
jupyter core     : 4.6.3
jupyter-notebook : 6.1.3
qtconsole        : not installed
ipython          : 5.8.0
ipykernel        : 5.3.4
jupyter client   : 6.1.6
jupyter lab      : 2.2.5
nbconvert        : 5.6.1
ipywidgets       : not installed
nbformat         : 5.0.7
traitlets        : 4.3.3
```
<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
        /home/*****/miniconda3/envs/jlab-test/bin
        /home/*****/miniconda3/condabin
        /home/*****/miniconda3/bin
        /usr/local/ncbi/sra-tools/bin
        /usr/lib64/qt-3.3/bin
        /usr/local/bin
        /bin
        /usr/bin
        /usr/local/sbin
        /usr/sbin
        /sbin
        /opt/ant/bin
        /home/*****/bin

sys.path:
        /home/*****/miniconda3/envs/jlab-test/bin
        /home/*****/miniconda3/envs/jlab-test/lib/python37.zip
        /home/*****/miniconda3/envs/jlab-test/lib/python3.7
        /home/*****/miniconda3/envs/jlab-test/lib/python3.7/lib-dynload
        /home/*****/miniconda3/envs/jlab-test/lib/python3.7/site-packages

sys.executable:
        /home/*****/miniconda3/envs/jlab-test/bin/python

sys.version:
        3.7.8 | packaged by conda-forge | (default, Jul 31 2020, 02:25:08)
        [GCC 7.5.0]

platform.platform():
        Linux-2.6.32-696.23.1.el6.x86_64-x86_64-with-centos-6.8-Final

which -a jupyter:
        /home/*****/miniconda3/envs/jlab-test/bin/jupyter
        /home/*****/miniconda3/bin/jupyter

pip list:
        Package                       Version
        ----------------------------- -------------------
        argon2-cffi                   20.1.0
        attrs                         19.3.0
        backports.functools-lru-cache 1.6.1
        bleach                        3.1.5
        brotlipy                      0.7.0
        certifi                       2020.6.20
        cffi                          1.14.1
        chardet                       3.0.4
        cryptography                  3.0
        decorator                     4.4.2
        defusedxml                    0.6.0
        entrypoints                   0.3
        idna                          2.10
        importlib-metadata            1.7.0
        ipykernel                     5.3.4
        ipython                       5.8.0
        ipython-genutils              0.2.0
        Jinja2                        2.11.2
        json5                         0.9.4
        jsonschema                    3.2.0
        jupyter-client                6.1.6
        jupyter-core                  4.6.3
        jupyterlab                    2.2.5
        jupyterlab-server             1.2.0
        MarkupSafe                    1.1.1
        mistune                       0.8.4
        nbconvert                     5.6.1
        nbformat                      5.0.7
        notebook                      6.1.3
        packaging                     20.4
        pandocfilters                 1.4.2
        pexpect                       4.8.0
        pickleshare                   0.7.5
        pip                           20.2.2
        prometheus-client             0.8.0
        prompt-toolkit                1.0.15
        ptyprocess                    0.6.0
        pycparser                     2.20
        Pygments                      2.6.1
        pyOpenSSL                     19.1.0
        pyparsing                     2.4.7
        pyrsistent                    0.16.0
        PySocks                       1.7.1
        python-dateutil               2.8.1
        pyzmq                         19.0.2
        requests                      2.24.0
        Send2Trash                    1.5.0
        setuptools                    49.6.0.post20200814
        simplegeneric                 0.8.1
        six                           1.15.0
        terminado                     0.8.3
        testpath                      0.4.4
        tornado                       6.0.4
        traitlets                     4.3.3
        urllib3                       1.25.10
        wcwidth                       0.2.5
        webencodings                  0.5.1
        wheel                         0.35.1
        zipp                          3.1.0

conda list:
        # packages in environment at /home/*****/miniconda3/envs/jlab-test:
        #
        # Name                    Version                   Build  Channel
        _libgcc_mutex             0.1                 conda_forge    conda-forge
        _openmp_mutex             4.5                       1_gnu    conda-forge
        argon2-cffi               20.1.0           py37h8f50634_1    conda-forge
        attrs                     19.3.0                     py_0    conda-forge
        backports                 1.0                        py_2    conda-forge
        backports.functools_lru_cache 1.6.1                      py_0    conda-forge
        bleach                    3.1.5              pyh9f0ad1d_0    conda-forge
        brotlipy                  0.7.0           py37h8f50634_1000    conda-forge
        ca-certificates           2020.6.20            hecda079_0    conda-forge
        certifi                   2020.6.20        py37hc8dfbb8_0    conda-forge
        cffi                      1.14.1           py37h2b28604_0    conda-forge
        chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
        cryptography              3.0              py37hb09aad4_0    conda-forge
        decorator                 4.4.2                      py_0    conda-forge
        defusedxml                0.6.0                      py_0    conda-forge
        entrypoints               0.3             py37hc8dfbb8_1001    conda-forge
        idna                      2.10               pyh9f0ad1d_0    conda-forge
        importlib-metadata        1.7.0            py37hc8dfbb8_0    conda-forge
        importlib_metadata        1.7.0                         0    conda-forge
        ipykernel                 5.3.4            py37h43977f1_0    conda-forge
        ipython                   5.8.0                    py37_1    conda-forge
        ipython_genutils          0.2.0                      py_1    conda-forge
        jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
        json5                     0.9.4              pyh9f0ad1d_0    conda-forge
        jsonschema                3.2.0            py37hc8dfbb8_1    conda-forge
        jupyter_client            6.1.6                      py_0    conda-forge
        jupyter_core              4.6.3            py37hc8dfbb8_1    conda-forge
        jupyterlab                2.2.5                      py_0    conda-forge
        jupyterlab_server         1.2.0                      py_0    conda-forge
        ld_impl_linux-64          2.34                 hc38a660_9    conda-forge
        libffi                    3.2.1             he1b5a44_1007    conda-forge
        libgcc-ng                 9.3.0               h24d8f2e_15    conda-forge
        libgomp                   9.3.0               h24d8f2e_15    conda-forge
        libsodium                 1.0.18               h516909a_0    conda-forge
        libstdcxx-ng              9.3.0               hdf63c60_15    conda-forge
        markupsafe                1.1.1            py37h8f50634_1    conda-forge
        mistune                   0.8.4           py37h8f50634_1001    conda-forge
        nbconvert                 5.6.1            py37hc8dfbb8_1    conda-forge
        nbformat                  5.0.7                      py_0    conda-forge
        ncurses                   6.2                  he1b5a44_1    conda-forge
        notebook                  6.1.3            py37hc8dfbb8_0    conda-forge
        openssl                   1.1.1g               h516909a_1    conda-forge
        packaging                 20.4               pyh9f0ad1d_0    conda-forge
        pandoc                    2.10.1               h516909a_0    conda-forge
        pandocfilters             1.4.2                      py_1    conda-forge
        pexpect                   4.8.0            py37hc8dfbb8_1    conda-forge
        pickleshare               0.7.5           py37hc8dfbb8_1001    conda-forge
        pip                       20.2.2                     py_0    conda-forge
        prometheus_client         0.8.0              pyh9f0ad1d_0    conda-forge
        prompt_toolkit            1.0.15                     py_1    conda-forge
        ptyprocess                0.6.0                   py_1001    conda-forge
        pycparser                 2.20               pyh9f0ad1d_2    conda-forge
        pygments                  2.6.1                      py_0    conda-forge
        pyopenssl                 19.1.0                     py_1    conda-forge
        pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
        pyrsistent                0.16.0           py37h8f50634_0    conda-forge
        pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
        python                    3.7.8           h6f2ec95_1_cpython    conda-forge
        python-dateutil           2.8.1                      py_0    conda-forge
        python_abi                3.7                     1_cp37m    conda-forge
        pyzmq                     19.0.2           py37hac76be4_0    conda-forge
        readline                  8.0                  he28a2e2_2    conda-forge
        requests                  2.24.0             pyh9f0ad1d_0    conda-forge
        send2trash                1.5.0                      py_0    conda-forge
        setuptools                49.6.0           py37hc8dfbb8_0    conda-forge
        simplegeneric             0.8.1                      py_1    conda-forge
        six                       1.15.0             pyh9f0ad1d_0    conda-forge
        sqlite                    3.33.0               h4cf870e_0    conda-forge
        terminado                 0.8.3            py37hc8dfbb8_1    conda-forge
        testpath                  0.4.4                      py_0    conda-forge
        tk                        8.6.10               hed695b0_0    conda-forge
        tornado                   6.0.4            py37h8f50634_1    conda-forge
        traitlets                 4.3.3            py37hc8dfbb8_1    conda-forge
        urllib3                   1.25.10                    py_0    conda-forge
        wcwidth                   0.2.5              pyh9f0ad1d_1    conda-forge
        webencodings              0.5.1                      py_1    conda-forge
        wheel                     0.35.1             pyh9f0ad1d_0    conda-forge
        xz                        5.2.5                h516909a_1    conda-forge
        zeromq                    4.3.2                he1b5a44_3    conda-forge
        zipp                      3.1.0                      py_0    conda-forge
        zlib                      1.2.11            h516909a_1007    conda-forge
</pre>
</details>

**Comment by potoo0** [comment]
3.0 still with this bug.

# Context
- Operating System and version: windows 10 20H2
- Browser and version: Chrome 87.0.4280.141
- JupyterLab version:

```bash
% jupyter --version
jupyter core     : 4.6.3
jupyter-notebook : 6.2.0
qtconsole        : not installed
ipython          : 7.10.0
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 3.0.5
nbconvert        : 6.0.2
ipywidgets       : 7.5.1
nbformat         : 5.0.7
traitlets        : 5.0.4
```

# Troubleshoot Screenshot:
![555](https://user-images.githubusercontent.com/34411681/105271191-0a5c1a00-5bd2-11eb-9d4d-05b62b12f5b7.gif)

**Comment by krassowski** [comment]
I can confirm this issue locally using version 3.0.12 and on binder using 3.0.7. It happens when the directory is changed, but the new listing is longer than the previous listing. I have run into a very similar issue myself when on this PR that also modifies listing: https://github.com/jupyterlab/jupyterlab-git/pull/885, but I cannot remember how I resolved it.

![repro](https://user-images.githubusercontent.com/5832902/112869032-9e57cf80-90b4-11eb-8f2c-2d0582abdcc2.gif)

**Comment by krassowski** [comment]
Looking at the current implementation, it is obvious in the hindsight that it will not work with listing items added afterwards (and only works in other folders _at all_ because the listing re-uses old items if available):

https://github.com/jupyterlab/jupyterlab/blob/879385b7cb9eba2a5d5975035b481f0c69022782/packages/filebrowser-extension/src/index.ts#L907-L927

There are two solutions I can imagine:
1. Use `MutationObserver` to detect addition of new elements with CSS class: `jp-DirListing-itemModified`
   - pros: can patched the 3.0 and 2.x branch
   - cons: requires a modern browser (IE11+, global support >97.5% https://caniuse.com/mutationobserver) - I don't think this is a problem, but I am not really sure what ranges of browsers JupyterLab wants to support
2. Add new `toggleColumn` method to the `DirListing.IRenderer` which will influence actual rendering (so that the extra DOM nodes are not shown when not needed).
   - pros: improves performance
   - cons: changes to the API - not the problem, but probably should go for 3.1

Any thoughts?

**Comment by salihkurucan** [comment]
And, is it possible to hide last modified column permanently?
